src/abstractdataframe/join.jl

```diff
@@ -290,6 +298,26 @@ function Base.join(df1::AbstractDataFrame,
 
     joiner = DataFrameJoiner(df1, df2, on)
 
+    # Check merge key validity
+    left_invalid = validate[1] ? any(nonunique(joiner.dfl, joiner.left_on)) : false
+    right_invalid = validate[2] ? any(nonunique(joiner.dfr, joiner.right_on)) : false
+
+    if left_invalid && !right_invalid
+        first_error = findfirst(nonunique(joiner.dfl, joiner.left_on))
+        throw(ArgumentError("Merge key(s) in df1 are not unique. " *
+                            "First duplicate at row $first_error"))
+    elseif !left_invalid && right_invalid
+        first_error = findfirst(nonunique(joiner.dfr, joiner.right_on))
+        throw(ArgumentError("Merge key(s) in df2 are not unique. " *
+                            "First duplicate at row $first_error"))
+    elseif left_invalid && right_invalid
```

Comment:
If you put this branch first, you can avoid the `!left_invalid` and `!right_invalid` checks.